### PR TITLE
[WIP] CuraPackage support

### PR DESCRIPTION
**Pull request by ChrisTerBeke**
The final implementation for CuraPackage. The PoC can be found at https://github.com/Ultimaker/libCharon/pull/2.

- [x] General cleanup
- [x] Some small fixes in typing and tests
- [x] Make an abstract implementation of OPC from which UFP and CuraPackage are children
- [x] Add convenience methods for interacting with Curapackage
  - [x] Materials
  - [x] Qualities
  - [x] Machines
  - [x] Plugins
- [x] Validate meta data when saving (package.json)

Todo:
- [ ] Get the exact required fields for package.json
- [ ] Reviews
- [ ] Fix merge conflicts

### CODE REVIEW

**Comment by ChrisTerBeke** [comment]
@nallath @Ghostkeeper can you have a look at this implementation? It does everything we currently need it to do from the server perspective (with the exception maybe of handling the material signature files?). We can extend it later to also add needed methods for reading everything into Cura.

File: Charon/filetypes/CuraPackage.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+            # Find which files to add.
+            for zip_item in zip_file.filelist:
+                if ignore_string.search(zip_item.filename):
+                    continue
+                paths_to_add.append(zip_item.filename)
+
+            # Validate required files.
+            for required_path in required_paths:
+                if required_path not in paths_to_add:
```
Comment (DanielSchiavini) [inline Charon/filetypes/CuraPackage.py]: The `in` operation is not efficient for lists, can we use a set for the `paths_to_add` variable?

File: docs/cura_package.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+### Adding a plugin
+A CuraPackage can contain one or more Cura plugins. These are stored under the `/files/plugins` directory in the package.
+A convenience method `addPlugin` is available in the library to sort this all out for you.
+
+```python
+plugin_root_path = os.path.abspath("MyCuraPlugin.zip")
+package.addPlugin(open(plugin_root_path, "rb").read(), plugin_id="MyCuraPlugin")
+```
+
+This code example will take a ZIP called `MyCuraPlugin/zip` that lives in your current working directory and adds it to the package as plugin.
```
Comment (DanielSchiavini) [inline docs/cura_package.md]: .zip

File: Charon/filetypes/CuraPackage.py (hunk 1 above)
Comment (nallath) [reply Charon/filetypes/CuraPackage.py]: The list will be relatively small, so i doubt you will even notice the optimisation.

File: Charon/filetypes/CuraPackage.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                stream = self.getStream("{}/{}".format(plugin_path_alias, path))
+                stream.write(zip_file.read(path))
+
+    ##  Export the package to bytes.
+    def toByteArray(self, offset: int = 0, count: int = -1) -> bytes:
+        self._validateMetadata()
+        return super().toByteArray(offset, count)
+
+    ##  Creates all the required content types for a .curapackage.
+    def _readContentTypes(self):
```
Comment (nallath) [inline Charon/filetypes/CuraPackage.py]: Missing return type

File: Charon/filetypes/CuraPackage.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        super()._readContentTypes()
+        if self.mode != OpenMode.ReadOnly:
+            self.addContentType(extension="xml.fdm_material", mime_type="application/x-ultimaker-material-profile")
+            self.addContentType(extension="xml.fdm_material.sig", mime_type="application/x-ultimaker-material-sig")
+            self.addContentType(extension="inst.cfg", mime_type="application/x-ultimaker-quality-profile")
+            self.addContentType(extension="def.json", mime_type="application/x-ultimaker-machine-definition")
+
+    ##  Validates if the package.json metadata file contains all the required keys
+    #   and if they are in the correct format.
+    def _validateMetadata(self):
```
Comment (nallath) [inline Charon/filetypes/CuraPackage.py]: missing return type

File: Charon/filetypes/OpenPackagingConvention.py [hunk 5]
```diff
@@ -0,0 +1,574 @@
...
+
+from collections import OrderedDict  # To specify the aliases in order.
+from io import BytesIO, TextIOWrapper  # For the type of input of openStream and to create binary output streams for getting metadata.
+import json  # The metadata format.
+import re  # To find the path aliases.
+from typing import Any, Dict, List, IO, Optional
+import xml.etree.ElementTree as ET  # For writing XML manifest files.
+import zipfile
+
+from ..FileInterface import FileInterface  # The interface we're implementing.
```
Comment (nallath) [inline Charon/filetypes/OpenPackagingConvention.py]: It's better to not use relative imports.

File: Charon/filetypes/OpenPackagingConvention.py [hunk 6]
```diff
@@ -0,0 +1,574 @@
...
+            metadata_per_file[file_name] = {}
+            for metadata_key in self.metadata:
+                if metadata_key.startswith(file_name + "/"):
+                    # Strip the prefix: "/a/b/c.stl/print_time" becomes just "print_time" about the file "/a/b/c.stl".
+                    metadata_per_file[file_name][metadata_key[len(file_name) + 1:]] = self.metadata[metadata_key]
+                    keys_left.remove(metadata_key)
+        # keys_left now contains only global metadata keys.
+
+        global_metadata = {key:self.metadata[key] for key in keys_left}
+        if len(global_metadata) > 0:
```
Comment (nallath) [inline Charon/filetypes/OpenPackagingConvention.py]: You don't need to use len here, you can just use if not global_metadata (empty dicts evaluate to False)

File: docs/cura_package.md (hunk 2 above)
Comment (ChrisTerBeke) [reply docs/cura_package.md]: Fixed

File: Charon/filetypes/CuraPackage.py (hunk 1 above)
Comment (ChrisTerBeke) [reply Charon/filetypes/CuraPackage.py]: Fixed

File: Charon/filetypes/OpenPackagingConvention.py [hunk 7]
```diff
@@ -0,0 +1,574 @@
...
+            self._writeMetadataToFile(global_metadata, self.global_metadata_file)
+            self.addRelation(self.global_metadata_file, self.opc_metadata_relationship_type)
+        for file_name, metadata in metadata_per_file.items():
+            if len(metadata) > 0:
```
Comment (nallath) [inline Charon/filetypes/OpenPackagingConvention.py]: Same as above.

File: Charon/filetypes/OpenPackagingConvention.py (hunk 5 above)
Comment (ChrisTerBeke) [reply Charon/filetypes/OpenPackagingConvention.py]: This was already there (OpenPackagingConvention is just renamed from the old UltimakerFormatPackage file). I'm seeing a lot of other relative imports as well, maybe that's needed to make it work correctly in Cura or the embedded system (python path settings)?

**Comment by ChrisTerBeke** [comment]
@nallath @awhiemstra I'm planning on fixing the conflicts this week. We have succcessfully integrated libCharon (and these PR changes) into our packaging workflow :)

**Comment by ChrisTerBeke** [comment]
Closing in favor of a new PR based on the refactor.

**Pull request closed without merging**